diff --git a/lib/index.js b/lib/index.js
index c9ad4c62ea9890311d406999e614df80e88a4273..b6cf10fc6b5d146ca9db9b396b4c614579bddec7 100644
--- a/lib/index.js
+++ b/lib/index.js
@@ -2038,10 +2038,16 @@ function setupFbtee({ hooks, translations }) {
 
 //#endregion
 //#region lib-tmp/createLocaleContext.tsx
+const {Context, useLocaleContext} = (() => {
+  'use client';
+
 const Context = createContext(null);
 function useLocaleContext() {
 	return use(Context);
 }
+
+return {Context, useLocaleContext}
+})()
 function setupLocaleContext({ availableLanguages, clientLocales, fallbackLocale = "en_US", gender = IntlVariations_default.GENDER_UNKNOWN, hooks, loadLocale, translations = { [fallbackLocale]: {} } }) {
 	const availableLocales = /* @__PURE__ */ new Map();
 	let currentLocale;
@@ -2093,6 +2099,7 @@ function setupLocaleContext({ availableLanguages, clientLocales, fallbackLocale
 	};
 }
 function createLocaleContext(props) {
+  'use client';
 	const { getLocale, setLocale } = setupLocaleContext(props);
 	return function LocaleContext({ children }) {
 		const [locale, _setLocale, localeChangeIsPending] = useActionState(async (previousLocale, newLocale) => {
